libs/x-content/src/main/java/org/elasticsearch/xcontent/SerializableString.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.xcontent;
+
+import java.util.function.Function;
+
+public class SerializableString {
+    private final String name;
+    private Object serializedString = null;
+
+    public static SerializableString of(String name) {
+        return new SerializableString(name);
+    }
+
+    private SerializableString(String name) {
+        this.name = name;
+    }
+
+    public <T> T computeIfAbsent(Class<T> clazz, Function<String, T> supplier) {
+        if (serializedString != null) {
+            assert serializedString.getClass().equals(clazz);
+            return clazz.cast(serializedString);
+        }
+        var serializedString = supplier.apply(name);
+        this.serializedString = serializedString;
+        return serializedString;
+    }
```

Comment:
Should we make initialization more deterministic? Also, given we know `T`, just cast using `(T)`?
```suggestion
    public <T> T computeIfAbsent(Class<T> clazz, Function<String, T> factory) {
        if (serializedString == null) {
            synchronized (this) {
                if (serializedString == null) {
                    return (T) (serializedString = factory.apply(name));
                }
            }
        }
        assert serializedString.getClass().equals(clazz);
        return (T) serializedString;
    }
```

Reply:
We discussed this a bit with Armin if you remember - the worst that can happen is that the String is serialized twice, and we "give out" different copies.
But then again using the double-checked locking pattern would make the synchronization block overhead negligible, so I'm OK with that!

Re: T, I feel it is cleaner to do a `Class.cast` instead of `(T)`, but I'm not rigid about it (unless using (T) gives us some IDE or compiler warning - if so, I'd stick with `Class.cast`)

Reply:
+1 to using clazz.cast

Reply:
I remembered that this pattern is safe in C# under some conditions, so I've looked up if it's safe in Java too (as Java has a weaker memory model), and it seems it's not: https://wiki.sei.cmu.edu/confluence/display/java/LCK10-J.+Use+a+correct+form+of+the+double-checked+locking+idiom
In particular I fear that the double-checking as it is written now may run into 2 problems:
1. lack of happens-before relationship between the first read at line 26 and the third read at line 34 means that it is possible for the third read (line 34) to obtain a stale null value (perhaps because its value was cached or reordered by the compiler), causing the method to return a null pointer.
2. return the newly created (non-null) value of the reference, but the object could be seen to be in an invalid or incorrect state - the JMM guarantees that objects are fully constructed before they become visible to any other thread only if they are immutable.

I think you can get rid of the first problem this way:
```
if (v != null) {
   return clazz.cast(v);
}
var v = this.serializedValue;      // Only un-synchronized read
synchronized (this) {
      v = serializedValue;          // In synchronized block, so this is safe
      if (v == null) {
         v = supplier.apply(value);
         this.serializedValue = v; 
      }
}
return clazz.cast(v);
```
But I also think this suffers from no.2. BTW, I believe my original code suffers from no.2 as well. 
Now, all parameters in jackson SerializableString are final or effectively final, so I think we are safe here, but we cannot assume it for every <T> -- can we?

The easy fix would be to make `serializedString` volatile, which was what we were hoping not to do. I'm calling @original-brownbear here to see if it's really important.

If it is, we need to find a way to make sure that threads always observe `serializedValue` to be either null or a fully constructed, valid object. Let me think about it a bit more :)

Reply:
Thanks for the fix on 1), you're lines are just a bit out of order.
Good point on 2, yes I think both our versions could theoretically suffer from that but should be fine for `SerializedString` (with `_value` being final). If we want to be safe, we could store serializedValue in an immutable helper (with a single final field), right?

Reply:
Yes, that's my understanding too. 
I think that introducing a Holder class that holds the newly created serializedValue in a final field should guarantee safe publication of the reference, at the cost of one more level of indirection.
But I'll wait for more expert people to chime in/confirm that.

Reply:
Yep, that should work: https://shipilev.net/blog/2014/safe-public-construction/
Pretty much identical (the if is reversed) but for the immutable helper/final wrapper:
```
private FinalWrapper wrappedSerializedValue;

public <T> T computeIfAbsent(Class<T> clazz, Function<String, T> factory) {
    var v = this.wrappedSerializedValue;
    if (v == null) {
      synchronized(this) {
        v = this.wrappedSerializedValue;
        if (v == null) {
          v = new FinalWrapper(supplier.apply(value));
          this.wrappedSerializedValue = v; 
        }
      }
    }
    return clazz.cast(v.instance);
  }

  private static class FinalWrapper {
    public final Object value;
    public FinalWrapper(Object value) {
      this.value = value;
    }
  }
```

But it also seems to gain little over volatile reads (that article has some good explanations and benchmarks)